### PR TITLE
Feature [RM61] Created and Implemented CharactersViewModel

### PR DESCRIPTION
**Pull request by swg99**
[https://github.com/orgs/novoda/projects/1#card-65878596](url)
The CharactersView now consumes a CharactersViewModel which is injected into the view from the ContentView. CharactersViewModel provides all the data for the CharactersView.

### CODE REVIEW

File: Rick-and-Morty/Rick And Morty/ViewModels/CharactersViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        
+        fetchCharacters()
+    }
+    
+    private func fetchCharacters() {
+        if characterType == .rick {
+            characters = ricks
+        } else {
+            characters = morties
+        }
```
Comment (MiWierzbinski) [inline Rick-and-Morty/Rick And Morty/ViewModels/CharactersViewModel.swift:23]: The function would be more testable with return value `private func fetchCharacters() -> [Character]`. 

Also for this length of function you could use 
```
characters = characterType == .rick ? ricks : morties
```

File: Rick-and-Morty/Rick And Morty/ViewModels/CharactersViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    }
+    
+    private func getTitle() -> String {
+        if characterType == .rick {
+            return "Ricks"
+        } else {
+            return "Morties"
+        }
+    }
```
Comment (MiWierzbinski) [inline Rick-and-Morty/Rick And Morty/ViewModels/CharactersViewModel.swift:32]: Similar here, 
```
return  characterType == .rick ? "Ricks" : "Morties"
```

File: Rick-and-Morty/Rick And Morty/ViewModels/CharactersViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+extension CharactersViewModel {
+    enum CharacterType {
+        case rick
+        case morty
+    }
+}
```
Comment (gbasile) [inline Rick-and-Morty/Rick And Morty/ViewModels/CharactersViewModel.swift:48]: Interesting: this is violating the Open Closed principle :P 
You would have to update the ViewModel if you want to add `Summer` or `Jerry`

Why don't you define a struct instead? 

```
struct CharacterViewState {
    let title: String
    let imagePosition: CharacterImagePosition
}
```

If you feed your ViewModel with an array of this model, you won't need to take decisions anymore inside your ViewModel ;)

File: Rick-and-Morty/Rick And Morty/ViewModels/CharactersViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    
+    init(characterType: CharacterType) {
+        self.characterType = characterType
+        self.title = getTitle()
+        self.imagePosition = getImagePosition()
+        
+        fetchCharacters()
+    }
+    
+    private func fetchCharacters() {
```
Comment (gbasile) [inline Rick-and-Morty/Rick And Morty/ViewModels/CharactersViewModel.swift:18]: This ViewModel has 2 responsibilities: 

- Feed the views with ViewStates
- Decide how to fetch the data

Can you imagine how you can move one of this responsibility out?

File: Rick-and-Morty/Rick And Morty/ViewModels/CharactersViewModel.swift (hunk 3 above)
Comment (swg99) [reply Rick-and-Morty/Rick And Morty/ViewModels/CharactersViewModel.swift:36]: I created a CharacterViewState and a CharactersViewState which holds an array of CharacterViewState's (naming could probably be better). 

`struct CharactersViewState: Identifiable {
    let id = UUID()
    let title: String
    let iconName: String
    let characterViewStates: [CharacterViewState]
}`

`struct CharacterViewState: Identifiable {
    let id = UUID()
    let name: String
    let description: String
    let shortDescription: String?
    let imageName: String
    let imagePosition: CharacterImagePosition
}`

A CharactersViewState is made for each CharacterType in the new ContentViewModel. I don't think there is a need for a CharactersViewModel anymore. Check out my new commit, it should make more sense.

File: Rick-and-Morty/Rick And Morty/ViewModels/CharactersViewModel.swift (hunk 4 above)
Comment (swg99) [reply Rick-and-Morty/Rick And Morty/ViewModels/CharactersViewModel.swift]: Moved this out into a CharactersManager class. The class needs to be modified if new CharacterTypes are added, have not been able to find a way around this.

File: Rick-and-Morty/Rick And Morty/Views/CharactersView.swift [hunk 5]
```diff
@@ -1,30 +1,28 @@
...
 
+/*
 struct CharactersView_Previews: PreviewProvider {
     static var previews: some View {
-        CharactersView(characters: ricks, title: "Ricks")
+        //CharactersView(viewModel: CharactersViewModel(characterType: .rick))
+        //CharactersView(viewModel: CharactersViewModel(characterType: .morty))
     }
 }
+*/
```
Comment (MiWierzbinski) [inline Rick-and-Morty/Rick And Morty/Views/CharactersView.swift:28]: This should be removed. there is no need for commented out code. You have a repo to check for old versions if you really need to :)

**Review by MiWierzbinski — APPROVED** [review]

**Pull request merged**